### ISSUE TITLE
TEE Plug-in for Eclipse (14.0.3): Missing libwebkitgtk dependency causes com.microsoft.vss.client.core.model.VssResourceNotFoundException: API resource location ada996bc-8c18-4193-b20c-cd41b13f5b4d is not registered on https://app.vssps.visualstudio.com/. Not Found

### ISSUE THREAD

**Issue by ramoas** [Microsoft/team-explorer-everywhere#85]
I am running Eclipse from the Spring Tool Suite 3.7.3 (Eclipse Luna SR2 (4.4.2) platform) on Ubuntu 14.04.

Since upgrading the TEE Plug-in for Eclipse to the 14.0.3 (14.0.3.201603291051) released version, I cannot seem to get past the below exception when clicking 'Finish' in the connection wizard after selecting the single TFVC project from the list in my VSTS account.  It's able to connect and query the list of available projects just fine.

I have had both this VSTS account and TFVC project for several years (circa 2012) -- since Visual Studio Online was still in Beta.

I have tried multiple troubleshooting steps: uninstalling / re-installing the plug-in, clearing out the ~/.microsoft folder, etc.

I have not seen this exception when using the command-line client (CLC) from 14.0.3 (14.0.3.201603291047), but perhaps I am simply not executing the same functionality as the connection wizard from within Eclipse.

For now, I have had to downgrade back to TEE 2013 (12.0.2.201409181809) in order for things to work.

I was not able to find and try any of the older versions of TEE 2015 (14.0.0, 14.0.1, 14.0.2) to determine when exactly the issue was introduced.

Here's the stack trace from the .metadata/.log file (looking at the code, the GUID ada996bc-8c18-4193-b20c-cd41b13f5b4d seems to be hard-coded in the method createSessionToken of DelegatedAuthorizationHttpClientBase):

com.microsoft.vss.client.core.model.VssResourceNotFoundException: API resource location ada996bc-8c18-4193-b20c-cd41b13f5b4d is not registered on https://app.vssps.visualstudio.com/. Not Found
        at com.microsoft.vss.client.core.VssHttpClientBase.createTarget(VssHttpClientBase.java:205)
        at com.microsoft.vss.client.core.VssHttpClientBase.createRequest(VssHttpClientBase.java:510)
        at com.microsoft.visualstudio.services.delegatedauthorization.DelegatedAuthorizationHttpClientBase.createSessionToken(DelegatedAuthorizationHttpClientBase.java:67)
        at com.microsoft.visualstudio.services.delegatedauthorization.DelegatedAuthorizationHttpClient.createAccountCodeAccessToken(DelegatedAuthorizationHttpClient.java:27)
        at com.microsoft.tfs.client.common.credentials.CredentialsHelper.createAccountCodeAccessToken(CredentialsHelper.java:62)
        at com.microsoft.tfs.client.common.ui.wizard.connectwizard.ConnectWizard.finishConnection(ConnectWizard.java:256)
        at com.microsoft.tfs.client.eclipse.ui.wizard.teamprojectwizard.EclipseTeamProjectWizard.doPerformFinish(EclipseTeamProjectWizard.java:172)
        at com.microsoft.tfs.client.common.ui.framework.wizard.ExtendedWizard.performFinish(ExtendedWizard.java:506)
        at org.eclipse.jface.wizard.WizardDialog.finishPressed(WizardDialog.java:853)
        at org.eclipse.jface.wizard.WizardDialog.buttonPressed(WizardDialog.java:438)
        at org.eclipse.jface.dialogs.Dialog$2.widgetSelected(Dialog.java:619)
        at org.eclipse.swt.widgets.TypedListener.handleEvent(TypedListener.java:248)
        at org.eclipse.swt.widgets.EventTable.sendEvent(EventTable.java:84)
        at org.eclipse.swt.widgets.Display.sendEvent(Display.java:4454)
        at org.eclipse.swt.widgets.Widget.sendEvent(Widget.java:1388)
        at org.eclipse.swt.widgets.Display.runDeferredEvents(Display.java:3799)
        at org.eclipse.swt.widgets.Display.readAndDispatch(Display.java:3409)
        at org.eclipse.jface.window.Window.runEventLoop(Window.java:832)
        at org.eclipse.jface.window.Window.open(Window.java:808)
        at com.microsoft.tfs.client.common.ui.teamexplorer.helpers.ConnectHelpers.connectToServer(ConnectHelpers.java:33)
        at com.microsoft.tfs.client.common.ui.controls.teamexplorer.TeamExplorerControl$ConnectToServerAction.run(TeamExplorerControl.java:677)
        at org.eclipse.jface.action.Action.runWithEvent(Action.java:519)
        at org.eclipse.jface.action.ActionContributionItem.handleWidgetSelection(ActionContributionItem.java:595)
        at org.eclipse.jface.action.ActionContributionItem.access$2(ActionContributionItem.java:511)
        at org.eclipse.jface.action.ActionContributionItem$5.handleEvent(ActionContributionItem.java:420)
        at org.eclipse.swt.widgets.EventTable.sendEvent(EventTable.java:84)
        at org.eclipse.swt.widgets.Display.sendEvent(Display.java:4454)
        at org.eclipse.swt.widgets.Widget.sendEvent(Widget.java:1388)
        at org.eclipse.swt.widgets.Display.runDeferredEvents(Display.java:3799)
        at org.eclipse.swt.widgets.Display.readAndDispatch(Display.java:3409)
        at org.eclipse.e4.ui.internal.workbench.swt.PartRenderingEngine$9.run(PartRenderingEngine.java:1151)
        at org.eclipse.core.databinding.observable.Realm.runWithDefault(Realm.java:332)
        at org.eclipse.e4.ui.internal.workbench.swt.PartRenderingEngine.run(PartRenderingEngine.java:1032)
        at org.eclipse.e4.ui.internal.workbench.E4Workbench.createAndRunUI(E4Workbench.java:148)
        at org.eclipse.ui.internal.Workbench$5.run(Workbench.java:636)
        at org.eclipse.core.databinding.observable.Realm.runWithDefault(Realm.java:332)
        at org.eclipse.ui.internal.Workbench.createAndRunWorkbench(Workbench.java:579)
        at org.eclipse.ui.PlatformUI.createAndRunWorkbench(PlatformUI.java:150)
        at org.eclipse.ui.internal.ide.application.IDEApplication.start(IDEApplication.java:135)
        at org.eclipse.equinox.internal.app.EclipseAppHandle.run(EclipseAppHandle.java:196)
        at org.eclipse.core.runtime.internal.adaptor.EclipseAppLauncher.runApplication(EclipseAppLauncher.java:134)
        at org.eclipse.core.runtime.internal.adaptor.EclipseAppLauncher.start(EclipseAppLauncher.java:104)
        at org.eclipse.core.runtime.adaptor.EclipseStarter.run(EclipseStarter.java:380)
        at org.eclipse.core.runtime.adaptor.EclipseStarter.run(EclipseStarter.java:235)
        at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
        at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
        at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
        at java.lang.reflect.Method.invoke(Method.java:498)
        at org.eclipse.equinox.launcher.Main.invokeFramework(Main.java:648)
        at org.eclipse.equinox.launcher.Main.basicRun(Main.java:603)
        at org.eclipse.equinox.launcher.Main.run(Main.java:1465)
        at org.eclipse.equinox.launcher.Main.main(Main.java:1438)
Caused by: com.microsoft.tfs.core.httpclient.HttpException: Not Found
        at com.microsoft.vss.client.core.VssHttpClientBase.loadLocations(VssHttpClientBase.java:187)
        at com.microsoft.vss.client.core.VssHttpClientBase.getLocation(VssHttpClientBase.java:134)
        at com.microsoft.vss.client.core.VssHttpClientBase.createTarget(VssHttpClientBase.java:203)

**Comment by DavidStaheli** [comment]
Hi @ramoas, thanks for the great detail.  I apologize for the trouble and time this cost you.  We are looking into the cause and possible workarounds and will get back to you quickly.

**Comment by ramoas** [comment]
Hi @DavidStaheli.  A Microsoft moderator on the Forum (https://social.msdn.microsoft.com/Forums/vstudio/en-US/4bbf5aa7-aea8-47be-a1ad-8a0d19c941dd/api-resource-location-is-not-registered-on-my-visualstudio-tfs-in-eclipse?forum=tee; another person started this thread, and I joined it after running into the same issue) made 14.0.1 available for me, and I was able to confirm that 14.0.1 is working for me as well.

Based on the UI, it definitely looks like the connection dialog / wizard / logic was changed at some point after 14.0.1.  If I get access to 14.0.2, I can test to further narrow down when the problem was introduced.

If you need any more logs (I had previously posted my teamexplorer-plugin / teamexplorer-soap-plugin logs on the above Microsoft forum) or need me to debug anything further on my side if you're having difficulty reproducing the issue, please let me know.

Right now, I can get by with using TEE 14.0.1 / TEE 12.0.2 / TEE-CLC, so feel free to focus on root cause.

**Comment by yacaovsnc** [comment]
Hi @ramoas, this error usually indicates a permission issue - the http client used to make the server call is not properly authenticated.  

As you have noticed we are making changes in this area and would you be willing to try the latest private build of TEE and see if it solves the issue?  My email address is yacao AT Microsoft dot com, and if you email me I can reply with a private build.  :)

Thanks!

**Comment by ramoas** [comment]
After extensive troubleshooting, including using a test application and a TEE 14.0.4 private build (from @yacaovsnc), the root cause of the error was a missing dependency on libwebkitgtk-1.0-0.  This library was simply not installed.  To correct this, given that I was on Ubuntu, all I had to do was run 'sudo apt-get install libwebkitgtk-1.0-0'.

Another key factor was that I was using Eclipse from Spring Tool Suite (STS) 3.7.3, but I had upgraded from STS 3.6.4.  Thus, unlike brand new installations of STS 3.7.3 which apparently use GTK 3 by default, my Eclipse instance was still apparently using GTK 2.  (NOTE: I did not have libwebkitgtk-3.0-0 installed either, so I would have had the same problem even if STS 3.7.3 were using GTK 3.)

The final factor that added to the troubleshooting complexity is that I had been using alternate credentials for my VSTS account with versions of TEE prior to 14.0.3, but alternate credentials are no longer supported with TEE 14.0.3+.  Only Microsoft accounts are supported for the initial login with TEE 14.0.3+.  I was not previously aware of this change in behavior. (NOTE: 14.0.3 of the Command-Line Client (CLC) still supports VSTS alternate credentials.)

Normally, to connect to a VSTS account, TEE 14.0.3 will open a browser-based dialog with the title "Sign in to <VSTSACCOUNT>.visualstudio.com".  The current text of this login page states "Sign in" and "Use your Microsoft account."  If you try to enter your VSTS alternate credentials and click "Sign in", you should get an error that states: "Please enter your phone number or your email address in the format someone@example.com."

If the libwebkitgtk dependency is missing like it was originally in my case, TEE 14.0.3 will fail to open the browser-based dialog and fallback to a dialog with the title "Enter Password" with the following text: "Authorization failure connecting to 'https://<VSTSACCOUNT>.visualstudio.com/TeamFoundation/Administration/v3.0/LocationService.asmx' (authenticating using credentials (Default Credentials))   Enter password for https://<VSTSACCOUNT.visualstudio.com/".  You should never see this dialog in TEE 14.0.3+; it's tantamount to failure, but you would never know this coming from an older version of TEE if you had always been using VSTS alternate credentials.

If you continue with that fallback dialog, your VSTS alternate credentials will gladly be accepted and used for basic authentication, which actually succeeds for some endpoints.  However, after clicking 'Finish' on the "Add Existing Team Project: Team Project Selection" dialog, basic authentication with the VSTS alternate credentials ultimately fails against one of the VSTS endpoints, resulting in the error in the title of this issue: "Error occurred in wizard completion.  More details are available in the Eclipse Platform log. API resource location ada996bc-8c18-4193-b20c-cd41b13f5b4d is not registered on https://app.vssps.visualstudio.com/.  Not Found".

Using the TEE 14.0.4 private build, we eventually realized that the critical SWT exception was simply not being logged from the login library (now fixed), which was keeping us in the dark.  In retrospect, we realized that with TEE 14.0.3, the corresponding teamexplorer-plugin log (from ~/.microsoft/Team\ Foundation/4.0/Logs/) shows the identical SWT exception for the ACSCredentialsDialog: "WARN  [main] (com.microsoft.tfs.client.common.ui.dialogs.connect.ACSCredentialsDialog) SWT Browser failed to load for ACSCredentialsDialog
org.eclipse.swt.SWTError: No more handles [Browser style SWT.MOZILLA and Java system property org.eclipse.swt.browser.DefaultType=mozilla are not supported with GTK 3 as XULRunner is not ported for GTK 3 yet]".

Once this SWT exception was identified, this led to the following Eclipse SWT FAQ: http://www.eclipse.org/swt/faq.php#browserwebkitgtk.  From this FAQ, we realized the libwebkitgtk dependency was missing, and once we realized STS 3.7.3 (upgraded from STS 3.6.4) was still using GTK 2, not GTK 3, we finally realized libwebkitgtk-1.0-0 was the correct version needed, not libwebkitgtk-3.0-0.

FWIW, I did test forcing STS 3.7.3 (upgraded from STS 3.6.4) to use GTK 3 after also installing libwebkitgtk-3.0-0.  However, I noticed some weird UI rendering and behavior with STS 3.7.3 (upgraded from STS 3.6.4) in GTK 3 mode; thus, I officially recommend installing libwebkitgtk-1.0-0 if you run into this problem and you have a similar STS / Eclipse configuration.

It was quite the journey.  Along the way, we discussed many other potential changes / improvements for which I will be creating separate issues.

Big thanks to @yacaovsnc (and @AlexRukhlin)!

**Comment by yacaovsnc** [comment]
Thanks @ramoas for your help in identify the missing dependencies and your suggestions!  

Just add one more details.  User can find out which version of GTK Eclipse (STS) is using by going into "Help -> Installation Details -> Configuration" and look for property "org.eclipse.swt.internal.gtk.version".  **If using GTK2, this property will be missing from the report..**

Thanks again!

**Comment by ramoas** [comment]
@yacaovsnc The existence of the property "org.eclipse.swt.internal.gtk.version" is an indicator of what GTK version you're running, but it's not a definitive guarantee.  For example, in my case, with Eclipse from STS 3.7.3 (upgraded from STS 3.6.4), when I forced use of GTK3 via the environment variable (SWT_GTK3=1), the property "org.eclipse.swt.internal.gtk.version" still does not exist.

I do not know what conditions trigger setting that property, but just because the property is missing, does not mean that you're using GTK2.

Besides the environment variable SWT_GTK3, another way that the GTK version can be set is via the flag --launcher.GTK_version (either as a command-line argument or an INI setting):

[The Eclipse runtime options](http://help.eclipse.org/mars/index.jsp?topic=%2Forg.eclipse.platform.doc.isv%2Freference%2Fmisc%2Fruntime-options.html)

**Comment by yacaovsnc** [comment]
Correct @ramoas, both SWT_GTK3 and --launcher.GTK_version are user set properties that can change the run environment.

The spirit of the comment was to aid the user to find out what version of GTK they are using, and I assume if you want to find out about that, you didn't set the properties yourself.  :-)

We are in the process of finishing device flow support which hopefully will mitigate the SWT requirement.

**Comment by m-mccool** [comment]
Hi
I am trying to use the Eclipse TFS Everywhere plugin on a Mac. I am experiencing the same problem. When is there going to be a release with this fix. Alternatively can you make the previous versions of the plugin available so I can connect to TFS.

**Comment by ramoas** [comment]
@m-mccool If you install the appropriate libwebkitgtk dependency, then you should be able to use TEE 14.0.3 with your Microsoft account.  Alternate VSTS credentials are no longer supported with TEE 14.0.3.

If you still want to use your alternate VSTS credentials, then you can downgrade to TEE 14.0.1 or TEE 12.0.2.  You can still download TEE 14.0.1 from this [Microsoft OneDrive share provided by a moderator on the Microsoft forum noted above](https://1drv.ms/f/s!AresBGZVYryjhRlWm-82-NF5HStg).  You can download TEE 12.0.2 directly from the [Microsoft Download Center](https://www.microsoft.com/en-us/download/details.aspx?id=40785).

**Comment by ramoas** [comment]
@m-mccool Since you're using Mac OS, not Linux, the exact cause of your problem might be different than what's described here.  I would suggest reviewing the Eclipse SWT FAQ for topics related to OS X: http://www.eclipse.org/swt/faq.php.  If you cannot figure it out, and the older versions do not work for you either, I'd suggest opening up a new Issue specifically for Mac OS.  I'd  start by including your Mac OS version, Eclipse version, Java version, TFS version (on-prem or VSTS), your Eclipse configuration (described above), etc.

**Comment by m-mccool** [comment]
HI I managed to install the 2013 version of TFS Everywhere (http://www.microsoft.com/en-gb/download/details.aspx?id=40785). Does this suggest to you that this is a bug for the Mac on the 2015 version. Is the 2013 version be using the same webkitgtk dependency?

**Comment by DavidStaheli** [comment]
Apologies for the trouble with this.  While investigation continues, the TEE 2015 **14.0.2** version has been made available for download again.  It may take a few hours for the changes to propagate and become visible.
https://www.microsoft.com/en-us/download/details.aspx?id=47727

**Comment by m-mccool** [comment]
Thanks @DavidStaheli , that worked a treat. Much appreciated

**Comment by AdautoRamalho** [comment]
GTK error still exists in 14.111.1 when running on Ubuntu 16 with Eclipse Neon. 
Also, I'd like to mention that I tried to enable credential cache and alternate credentials login, but both where ignored by the plugin. The clc worked like a charm, but the plugin appear to ignore those configurations. 
Also installed the dependency library as mentioned, and it does not work on ubuntu 16. Not sure of what it is, but there is something to do with XULRunner not working with GTK 3 too.
Tried to force Eclipse to run on GTK2 instead of 3 using eclipse.ini, also didn't work.

Not sure if there is a work around this, got the 14.0.2 up and it works fine just as mentioned before.
